gum/arch-arm64/gumarm64relocator.c

```diff
@@ -346,58 +346,131 @@ gum_arm64_relocator_can_relocate (gpointer address,
     csh capstone;
     cs_err err;
     cs_insn * insn;
-    size_t count, i;
-    gboolean eoi;
+    GHashTable * targets_to_check, * checked_targets;
+    guint8 * code_cur, * address_cur;
+    guint code_size;
+    GHashTableIter iter;
+    gpointer key, value;
 
     err = cs_open (CS_ARCH_ARM64, CS_MODE_LITTLE_ENDIAN, &capstone);
     g_assert_cmpint (err, == , CS_ERR_OK);
     err = cs_option (capstone, CS_OPT_DETAIL, CS_OPT_ON);
     g_assert_cmpint (err, ==, CS_ERR_OK);
 
-    count = cs_disasm (capstone, rl.input_cur, 1024, rl.input_pc, 0, &insn);
-    g_assert (insn != NULL);
+    insn = cs_malloc(capstone);
+    targets_to_check = g_hash_table_new (NULL, NULL);
+    checked_targets = g_hash_table_new (NULL, NULL);
 
-    eoi = FALSE;
-    for (i = 0; i != count && !eoi; i++)
+    code_cur = rl.input_cur;
+    address_cur = rl.input_pc;
+    code_size = 1024;
+    gboolean tragic_exit = false;
+    while (true)
     {
-      cs_arm64 * d = &insn[i].detail->arm64;
+      gboolean go_ahead = true;
+      g_hash_table_add (checked_targets, code_cur);
+      while (go_ahead && cs_disasm_iter (capstone, &code_cur, &code_size,
+          &address_cur, insn))
+      {
+        cs_arm64 * d = &insn->detail->arm64;
+        switch (insn->id)
+        {
+          case ARM64_INS_B:
+          {
+            cs_arm64_op * op = &d->operands[0];
+            g_assert (op->type == ARM64_OP_IMM);
+            if (!g_hash_table_contains(checked_targets, op->imm))
+              g_hash_table_add (targets_to_check, op->imm);
+            go_ahead = d->cc != ARM64_CC_INVALID && d->cc != ARM64_CC_AL &&
+                d->cc != ARM64_CC_NV;
+            break;
+          }
+          case ARM64_INS_CBZ:
+          case ARM64_INS_CBNZ:
+          {
+            cs_arm64_op * op = &d->operands[1];
+            g_assert (op->type == ARM64_OP_IMM);
+            if (!g_hash_table_contains(checked_targets, op->imm))
+              g_hash_table_add (targets_to_check, op->imm);
+            break;
+          }
+          case ARM64_INS_TBZ:
+          case ARM64_INS_TBNZ:
+          {
+            cs_arm64_op * op = &d->operands[2];
+            g_assert (op->type == ARM64_OP_IMM);
+            if (!g_hash_table_contains(checked_targets, op->imm))
+              g_hash_table_add (targets_to_check, op->imm);
+            break;
+          }
+          case ARM64_INS_RET:
+          {
+            go_ahead = false;
+            break;
+          }
+          case ARM64_INS_BR:
+          {
+            go_ahead = false;
+            tragic_exit = true;
+            /*
+             * tragic exit: we stop and reduce to the safest case because
+             * we can not know where the br can jump. Optimization is possible,
+             * by guessing the register value from previouos instruction.
+             */
```

Comment:
Why is this a tragic exit? There may still be other branches that we encountered previously, right?